### PR TITLE
Epoch transition during last slot of epoch

### PR DESCRIPTION
**Pull request by djrtwo**
* Addresses @mkalinin's comment here https://github.com/ethereum/eth2.0-specs/issues/267#issuecomment-446139113
* Also solves #352 issue 

There are a number of bugs and undefined behaviors floating around the boundaries of epochs wrt shuffling. Putting this up for review. I need to take a look at the implications to other slot-based helpers before we merge.

Note: Does not solve #92 but is related

### CODE REVIEW

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -1578,34 +1578,37 @@ For each `exit` in `block.body.exits`:
 
 ## Per-epoch processing
 
-The steps below happen when `state.slot % EPOCH_LENGTH == 0`.
+The steps below happen when `state.slot % EPOCH_LENGTH == EPOCH_LENGTH - 1`.
 
 ### Helpers
 
+* Let `current_epoch_start_slot = state.slot - EPOCH_LENGTH + 1`.
+* Let `previous_epoch_start_slot = state.slot - 2 * EPOCH_LENGTH + 1`.
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: It seems now `state.slot=63` is the first valid slot number to perform per-epoch transtion? When `state.slot=63`, `previous_epoch_start_slot = state.slot - 2 * EPOCH_LENGTH + 1 = 63 - 2 * 64 + 1 = -64`.
1. I think we somehow assume that "slot number is an unsigned 64-bit integer"? In this special case, `-64` may not conform "the previous epoch start slot" context.
2. Might related to: https://github.com/ethereum/eth2.0-specs/issues/192

**Comment by JustinDrake** [comment]
A few thoughts:

* Avoiding unnecessary `+ 1` and `- 1` feels like win to avoid off-by-one bugs and keep things clean.
* To minimise the `+ 1` and `- 1` we have some flexibility regarding the placement of the processing logic, as well as reworking the genesis block (e.g. make the block at slot 0 be a real block, as opposed to the block at slot 1).

File: specs/core/0_beacon-chain.md (hunk 1 above)
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: I think putting a min on this value of 0 actually makes the logic work.

**Comment by djrtwo** [comment]
@JustinDrake We can avoid the +1 -1 logic if we have the epoch transition at the _start_ of the next slot. This will generally make the numbers cleaner but does not 100% address #352. We have the necessary data to get the proposer from the start of the next epoch (v-set and seed) but the `get_crosslink_committees_at_slot` and `get_beacon_proposer_index` helpers do not currently allow for retrieving the shufflings until after the state transition is performed.

I think we can modify/add shuffling helpers to address this. 
I'll pick it up in the morning

**Comment by djrtwo** [comment]
Let me know your thoughts on inserting the epoch transition into the _start_ of the `slot % EPOCH_LENGTH == 0` slots.

So pretty much:
* increment state.slot
* if start of epoch, transition
* do per-slot
* do per_block

**Comment by JustinDrake** [comment]
> Let me know your thoughts on inserting the epoch transition into the start of the `slot % EPOCH_LENGTH == 0` slots.

If you feel that's clean then great :) I understand this PR is somewhat of a blocker, so feel free to move fast on it.

> * if start of epoch, transition
> * do per-slot

Having the per-epoch and per-slot right next to each other is nice because per-epoch is a special case of per-slot.

**Comment by djrtwo** [comment]
closing in favor of #480

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -1656,14 +1659,14 @@ If `state.slot % ETH1_DATA_VOTING_PERIOD == 0`:
...
+* Set `state.justification_bitfield |= 1` and `state.justified_slot = current_epoch_start_slot` if `3 * current_epoch_boundary_attesting_balance >= 2 * total_balance`.
 
 Set `state.finalized_slot = state.previous_justified_slot` if any of the following are true:
 
-* `state.previous_justified_slot == state.slot - 2 * EPOCH_LENGTH and state.justification_bitfield % 4 == 3`
-* `state.previous_justified_slot == state.slot - 3 * EPOCH_LENGTH and state.justification_bitfield % 8 == 7`
-* `state.previous_justified_slot == state.slot - 4 * EPOCH_LENGTH and state.justification_bitfield % 16 in (15, 14)`
+* `state.previous_justified_slot == previous_epoch_start_slot and state.justification_bitfield % 4 == 3`
+* `state.previous_justified_slot == previous_epoch_start_slot - 1 * EPOCH_LENGTH and state.justification_bitfield % 8 == 7`
+* `state.previous_justified_slot == previous_epoch_start_slot - 2 * EPOCH_LENGTH and state.justification_bitfield % 16 in (15, 14)`
```
Comment (vbuterin) [inline specs/core/0_beacon-chain.md]: Weren't we going to fix https://github.com/ethereum/eth2.0-specs/issues/410 at the same time? Or did you want to get this PR merged first and then do a separate one for that?

Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: Closed this PR. I think https://github.com/ethereum/eth2.0-specs/pull/480 is the way to go. I will fix #410 when we decide which of the two approaches to go with.

#480 does the epoch transition at the _start_ of slots (before block processing) at which `slot % EPOCH_LENGTH == 0`. this avoids all the nasty -1/+1 logic found in this PR.

**Pull request closed without merging**